sklearn/decomposition/_base.py

```diff
@@ -57,26 +62,33 @@ def get_precision(self):
         precision : array, shape=(n_features, n_features)
             Estimated precision of data.
         """
+        xp, _ = get_namespace(self.components_)
+
         n_features = self.components_.shape[1]
 
         # handle corner cases first
         if self.n_components_ == 0:
-            return np.eye(n_features) / self.noise_variance_
+            return xp.eye(n_features) / self.noise_variance_
 
-        if np.isclose(self.noise_variance_, 0.0, atol=0.0):
-            return linalg.inv(self.get_covariance())
+        if xp.isclose(
+            self.noise_variance_, xp.zeros_like(self.noise_variance_), atol=0.0
+        ):
+            return xp.linalg.inv(self.get_covariance())
 
         # Get precision using matrix inversion lemma
         components_ = self.components_
         exp_var = self.explained_variance_
         if self.whiten:
-            components_ = components_ * np.sqrt(exp_var[:, np.newaxis])
-        exp_var_diff = np.maximum(exp_var - self.noise_variance_, 0.0)
-        precision = np.dot(components_, components_.T) / self.noise_variance_
-        precision.flat[:: len(precision) + 1] += 1.0 / exp_var_diff
-        precision = np.dot(components_.T, np.dot(linalg.inv(precision), components_))
+            components_ = components_ * xp.sqrt(exp_var[:, np.newaxis])
+        exp_var_diff = xp.maximum(
+            exp_var - self.noise_variance_, xp.zeros_like(exp_var)
+        )
+        precision = components_ @ components_.T / self.noise_variance_
+        # TODO use views instead?
+        precision.reshape(-1)[:: len(precision) + 1] += 1.0 / exp_var_diff
+        precision = components_.T @ xp.linalg.inv(precision) @ components_
         precision /= -(self.noise_variance_**2)
-        precision.flat[:: len(precision) + 1] += 1.0 / self.noise_variance_
+        precision.reshape(-1)[:: len(precision) + 1] += 1.0 / self.noise_variance_
```

Comment:
reshape is a function in the array API spec (this is the sort of thing you would catch by testing against numpy.array_api).

Reply:
This code is trying to alter the array in place. I think this can work with `xp.reshape(..., copy=False)`, but `copy=False` is not accepted in `numpy.array_api` yet: https://github.com/numpy/numpy/issues/23410.

Reply:
Oh I see. That needs to be fixed upstream. There are still some things missing from numpy.array_api unfortunately (including all v2022 stuff). Perhaps we should prioritize this for the next NumPy release (when is that again?). CC @rgommers

Reply:
We could also add a simple `reshape` helper to array_api_compat if the release cadence for fixing numpy.array_api is too slow here.

Reply:
> Perhaps we should prioritize this for the next NumPy release (when is that again?). CC @rgommers

Yes, we should. The release split will happen pretty soon - I haven't seen a release schedule, but I suspect it'll be the second half of May.

Reply:
+1 but as scikit-learn needs compat with older numpy versions we need the compat workaround suggested in https://github.com/scikit-learn/scikit-learn/pull/26315#discussion_r1186082492 anyway in scikit-learn for the time being.